### PR TITLE
Introduce STM32 RNG driver and integration in platform stm32mp1

### PR DESCRIPTION
**Pull request by etienne-lms**

### CODE REVIEW

File: core/drivers/stm32_rng.c [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		timeout_ref = timeout_init_us(RNG_TIMEOUT_US);
+		do {
+			uint32_t status = read32(rng_base + RNG_SR);
+
+			if ((status & (RNG_SR_SECS | RNG_SR_SEIS)) != 0U) {
+				size_t i;
+
+				/* Recommended by the SoC reference manual */
+				io_mask32(rng_base + RNG_SR, 0, RNG_SR_SEIS);
+				dmb();
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: Why is this needed and why a `dmb()`?

Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: mmh.. i'm not sure. Register `DR` shall be read 3 times to discard the FIFO content while `SR[SEIS]` (read below) is 0. IO-DEV memory is mapped without strong ordering. And it failed to test this case. I am removing them and check internally if required.

**Comment by jenswi-linaro** [comment]
Perhaps we need to investigate if we should change the way we're mapping device memory. I can't recall why we're using "strongly-ordered" instead of "device".

File: core/drivers/stm32_rng.c [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+#define RNG_TIMEOUT_US		1000
+
+struct stm32_rng_instance {
+	struct io_pa_va base;
+	unsigned long clock;
+};
+
+static struct stm32_rng_instance *stm32_rng;
+
+int stm32_rng_read_raw(uintptr_t rng_base, uint8_t *out, size_t size)
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: It's a bit hard to tell what's really going on in this function. The read from `rng_base + RNG_DR` is quite obvious but that rest is a bit of a mystery. Some more comments, especially on the inner loops would be welcome. If there's a public specification a link or at least something specific to search for to get more information that would be nice.

Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: ok, i'll add.

File: core/drivers/stm32_rng.c [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+				for (i = 12; i != 0; i--)
+					(void)read32(rng_base + RNG_DR);
+
+				if (read32(rng_base + RNG_SR) & RNG_SR_SEIS)
+					panic("RNG noise");
+			}
+
+			if (timeout_elapsed(timeout_ref)) {
+				/* Timeout elapsed but RNG may now be ready */
+				if (read32(rng_base + RNG_SR) & RNG_SR_DRDY)
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: A test was just performed, testing again here is redundant.

File: core/drivers/stm32_rng.c [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		/* RNG is ready: read data by 32bit word, up to 4 words */
+		for (count = 0; count < 4; count++) {
+			uint32_t data32 = read32(rng_base + RNG_DR);
+
+			memcpy(buf, &data32, MIN(len, sizeof(uint32_t)));
+			buf += MIN(len, sizeof(uint32_t));
+			len -= MIN(len, sizeof(uint32_t));
+
+			if (!len)
+				break;
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: We're done, why not just return success?

File: core/drivers/stm32_rng.c [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+	while (len) {
+		timeout_ref = timeout_init_us(RNG_TIMEOUT_US);
+
+		/* Wait RNG is ready, a.k.a has produced random samples */
+		do {
+			uint32_t status = read32(rng_base + RNG_SR);
+
+			/* Noise source error: try to conceal with the error */
+			if ((status & (RNG_SR_SECS | RNG_SR_SEIS)) != 0U) {
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: the `!= 0` test is a bit redundant here.

What if `(status & RNG_SR_DRDY)` here? Would that mean that we're ready to read already? If so, how about breaking the loop here instead of at the end with the `while(...)` test?

File: core/drivers/stm32_rng.c [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+				size_t i;
+
+				/*
+				 * Recommended by the SoC reference manual to
+				 * reconcile from noise source errors:
+				 * deassert SR[SEIS], empty 3 32bit words and
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: I guess this means reading 3 times, not 12 times as it's done below.

File: core/drivers/stm32_rng.c [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+int stm32_rng_read_raw(uintptr_t rng_base, uint8_t *out, size_t size)
+{
+	uint8_t *buf = out;
+	size_t len = size;
+	uint64_t timeout_ref = 0;
+	int count = 0;
+
+	if (!(read32(rng_base + RNG_CR) & RNG_CR_RNGEN))
+		write32(RNG_CR_RNGEN | RNG_CR_CED, rng_base + RNG_CR);
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: What does this do? Enable RNG generation for a few words or "forever"?

File: core/drivers/stm32_rng.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: I meant... i'll add few comment. The spec is not yet public but will soon be. Once it is, each driver will get its reference in its header comment, sure.

File: core/drivers/stm32_rng.c (hunk 7 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: I'll add a comment. Indeed this enables forever. I will add a disable sequence once data are read.

File: core/drivers/stm32_rng.c (hunk 5 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: ok for removing `!= 0`. 

If one of SR[SECS] or SR[SEIS] is set, data shall be discarded whatever SR[DRDY].

File: core/drivers/stm32_rng.c (hunk 6 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: My fault. It is 12 32bit words. My comment is wrong. Thanks.

File: core/drivers/stm32_rng.c (hunk 3 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: I've added that in case timeout elapsed by to TEE thread be suspended since last read of register SR.

File: core/drivers/stm32_rng.c [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+				if (read32(rng_base + RNG_SR) & RNG_SR_SEIS)
+					panic("RNG noise");
+			}
+
+			if (timeout_elapsed(timeout_ref)) {
+				/* Timeout elapsed but RNG may now be ready */
+				if (read32(rng_base + RNG_SR) & RNG_SR_DRDY)
+					break;
+
+				return 1;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: yes, you're right. Already talked about such error core handling. I'll move to `TEE_Result`. 
Thanks.

File: core/drivers/stm32_rng.c (hunk 4 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: True.

File: core/drivers/stm32_rng.c [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+{
+	void *fdt = NULL;
+	struct dt_node_info dt_info = { 0 };
+	int node = -1;
+
+	fdt = get_embedded_dt();
+	if (!fdt)
+		panic();
+
+	while (1) {
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: ok

File: core/drivers/stm32_rng.c [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+		stm32_rng = calloc(1, sizeof(*stm32_rng));
+		if (!stm32_rng)
+			panic();
+
+		assert(dt_info.clock != DT_INFO_INVALID_CLOCK &&
+		       dt_info.reg != DT_INFO_INVALID_REG);
+
+		stm32_rng->base.pa = dt_info.reg;
+		stm32_rng->clock = (unsigned long)dt_info.clock;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: We're losing the sign here hence the cast (`int` => `unsigned long`), no?

File: core/drivers/stm32_rng.c (hunk 5 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: Indeed the loop implementation is not good: it should always check SR[SECS|SEIS] before relying on SR[DRDY] value.

File: core/drivers/stm32_rng.c (hunk 3 above)
Comment (jenswi-linaro) [reply core/drivers/stm32_rng.c]: I see, since we're running with non-secure interrupts unmasked we can be suspended in this loop. If that happens we're more or less guaranteed to have a timeout which doesn't occur due to the hardware misbehaving. Please add something along that to the comment because it's not obvious.

File: core/drivers/stm32_rng.c [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+				if (read32(rng_base + RNG_SR) & RNG_SR_SEIS)
+					panic("RNG noise");
+			}
+
+			if (timeout_elapsed(timeout_ref)) {
+				/* Timeout elapsed but RNG may now be ready */
+				if (read32(rng_base + RNG_SR) & RNG_SR_DRDY)
+					break;
+
+				return TEE_ERROR_SECURITY;
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: We've leaving the RNG running here.

File: core/drivers/stm32_rng.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/drivers/stm32_rng.c]: As far as I know casting makes no difference here.

File: core/drivers/stm32_rng.c [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+			}
+
+		} while (!(read32(rng_base + RNG_SR) & RNG_SR_DRDY));
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: In some answer you said:
*Indeed the loop implementation is not good: it should always check SR[SECS|SEIS] before relying on SR[DRDY] value.*
I guess we need to check SR[SECS|SEIS] first.

File: core/drivers/stm32_rng.c (hunk 3 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: ok.

File: core/drivers/stm32_rng.c (hunk 12 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: Indeed. I have a fixup I am testing before pushing to this P-R.

File: core/drivers/stm32_rng.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: It is GCC that complains when losing sign without an explicit statement as a cast.

[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/drivers/stm32_rng.c]: Fair enough, it has a purpose.

File: core/drivers/stm32_rng.c [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+		write32(RNG_CR_RNGEN | RNG_CR_CED, rng_base + RNG_CR);
+
+	while (len) {
+		uint64_t timeout_ref = timeout_init_us(RNG_TIMEOUT_US);
+
+		/* Wait RNG has produced random well seeded samples */
+		while (!timeout_elapsed(timeout_ref)) {
+			if (!conceal_seed_error(rng_base) &&
+			    read32(rng_base + RNG_SR) & RNG_SR_DRDY)
+				break;
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: This will only break the inner loop. How about adding an `out:` label below, and set `rc` here as needed and `goto out`?

File: core/drivers/stm32_rng.c [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+		if (read32(rng_base + RNG_SR) & RNG_SR_SEIS)
+			panic("RNG noise");
+
+		return true;
+	}
+
+	return false;
+}
+
+TEE_Result stm32_rng_read_raw(uintptr_t rng_base, uint8_t *out, size_t size)
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: It just occurred to me that this function isn't protected from multiple calls in parallel. Is it the callers responsibility to protect against that?

File: core/drivers/stm32_rng.c [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+	while (len) {
+		uint64_t timeout_ref = timeout_init_us(RNG_TIMEOUT_US);
+
+		/* Wait RNG has produced random well seeded samples */
+		while (!timeout_elapsed(timeout_ref)) {
+			if (!conceal_seed_error(rng_base) &&
+			    read32(rng_base + RNG_SR) & RNG_SR_DRDY)
+				break;
+		}
+		if (conceal_seed_error(rng_base))
```
Comment (jenswi-linaro) [inline core/drivers/stm32_rng.c]: Why do we care if a seed error was concealed or not? Regardless it's OK now or we would have panicked.

File: core/drivers/stm32_rng.c (hunk 14 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: True, i should add a lock. Will do.

File: core/drivers/stm32_rng.c (hunk 13 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: This is intended to break the inner loop and move to the sequence below:
- conceal seed error, if occurrence, wait again for samples.
- test if data are ready, if not, then it's an effective timeout.

The current sequence allows to conceal with seed errors without waiting for the timeout to do so.

File: core/drivers/stm32_rng.c (hunk 4 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: This now breaks the loop to reach the function's epilogue: disable RNG and return with status.
Maybe `goto out;` would be more explicit.

File: core/drivers/stm32_rng.c (hunk 15 above)
Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: If a seed error occurred, we can still attempt to read samples, we should not go straight to a timeout condition, hence the conditioned `continue;` instruction here.

File: core/drivers/stm32_rng.c (hunk 13 above)
Comment (jenswi-linaro) [reply core/drivers/stm32_rng.c]: This is confusing. So if there's a conceal seed error we test again until it has vanished and data is ready. The first thing we do if data is ready is to test for conceal seed error and check if data is ready. If there's a timeout we check if the there's a conceal error, if so we restart the timeout and loop again.

The only way to get a timeout is if there isn't a conceal seed error but data isn't ready.

Is this a correct summary of intended behavior?

Comment (etienne-lms) [reply core/drivers/stm32_rng.c]: hmmm... i'm confused myself :(
I start adding some mutex for RNG accesses, using spinlocks. This makes the loop uninterruptible, hence simplifies the timeout detection part. I will push an update once I have something.

I plan to make the `stm32_rng_read_raw()`  accessing RNG by 16 bytes data burst in atomic context, and `stm32_rng_read()` to loop,  locking/unlocking RNG for each 16byte max burst.

**Comment by etienne-lms** [comment]
@jenswi-linaro, would you be ok if I squash the current fixup commits and rebase the whole, leaving the loop as-is so we can focus on this loop and TEE scheduling?

**Comment by jenswi-linaro** [comment]
Fine with me.

**Comment by etienne-lms** [comment]
I squashed to fixup commits with 1st one and rebase. On this, 2 added commits to resume review of the RNG loop.

First added commit is already proposed in https://github.com/OP-TEE/optee_os/pull/2806.
Then a fixup commit with a new scheme for the sequence: an uninterruptible sequence to read 16byte burst, and a interruptible sequence calling the former.

**Comment by etienne-lms** [comment]
Rebased to fix merge conflicts and upgrade some internal APIs (`io_read32()`, `register_phys_mem_pgdir()`).

The commit for reference counting is introduced back, and located as 1st commit of the series.

**Comment by etienne-lms** [comment]
Forced pushed an updated series to fix a issue reported by checkpatch.

**Comment by etienne-lms** [comment]
@jenswi-linaro, could you have a look at latest proposal?

File: core/drivers/stm32_rng.c [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+		memset(out, 0, size);
+
+	return rc;
+}
+
+#ifdef CFG_EMBED_DTB
+static TEE_Result stm32_rng_init(void)
+{
+	void *fdt = NULL;
+	struct dt_node_info dt_info = { 0 };
```
Comment (etienne-lms) [inline core/drivers/stm32_rng.c]: will replace with `memset()`.

**Comment by etienne-lms** [comment]
Patch "stm32mp1: fix RCC interface register mapping registering" removed.
Fixup commits squashed.
Tag applied.
